### DISCUSSION TITLE
Issue in GetScanResultClientRequestHandlerExt function

### DISCUSSION THREAD

**Discussion by PruthviIntwoC** [microsoft/Dynamics365Commerce.Solutions#19] [Q&A] [answered]
Hi myself pruthvi i am struck here can anyone help me out, detailed explaination is as below
I am trying to scan the barcode from Peripheral Simulator, and i am using the standard request handler i.e GetScanResultClientRequestHandlerExt. but i am unable to remove some characters to formulate a new barcode to the numpad.  
Here my output should be example: when i scan a item which has barcode (13 numbers) and epc code 24 number separated by comma(,) i just want barcode rest i need to remove but in the standard GetScanResultClientRequestHandlerExt function we can't able to remove the substring?
Can anyone answer my question?
![code](https://user-images.githubusercontent.com/119489652/204799200-aceddeb0-d780-4e81-9cb1-0e872624c5f3.png)
![UISimulator](https://user-images.githubusercontent.com/119489652/204799216-fb4beaf8-6305-4102-a9f3-aac31f16d8aa.png)

**Comment by madyke** [comment]
Hi Pruthvi, I tested this sample yesterday and it seems to be working fine for me. What issue are you seeing exactly?

One thing to check would be that when you're testing it, make sure you have focus on the numpad before using Peripheral Simulator to scan the barcode. That's when the `GetScanResultClientRequestHandlerExt` request handler will be called, where you can modify the `scanText` to get the substring. If you're just trying to scan from the POS homepage, without putting focus on the numpad, then that request handler won't be used and it will default to a product search.

**Reply by PruthviIntwoC** [reply]
Hi madyke, Thanks for the reply 
I have scanning while on focus numpad but it is not, even i tried manually enter the barcode value in the numpad it is going to the 
GetScanResultClientRequestHandlerExt handler but it is not getting substring 
can you please send me the screenshots which you have tried in the handler so that i can replicate and try

**Reply by madyke** [reply]
I haven't changed anything in the sample extension handler, other than uncommenting the line of code where it gets the substring:
![image](https://user-images.githubusercontent.com/13256055/207996161-df13bc87-5c25-4be3-aaa7-875c4cc2ec99.png)

Here's the walkthrough I'm doing to test the extension request handler, where I confirm that the extension has loaded, go the transaction page, focus on the numpad, and then use Peripheral Simulator to scan the barcode. In the extension request handler, the full `scanText` comes in as part of the request, where I'm then able to get the substring of just the first 13 characters which I pass to the default request hander execution:
![GetScanResultClientRequest](https://user-images.githubusercontent.com/13256055/207996256-1e731621-2462-4a6f-a372-c885bd83c8aa.gif)

**Reply by madyke** [reply]
I also discussed this with Luke offline, he noticed from your original screenshots that you're using the Retail SDK, which has been deprecated and should not be used for any new Commerce extension development. Switching to the Commerce SDK may resolve your issue if you're still having problems.

**Comment by PruthviIntwoC** [comment, marked as answer]
Hi Madyke, thanks for the reply please confirm that you are using commerce SDK or Retail SDK? The above overview you have shared is from RetailSDK or Commerce SDK. Also confirm once execution is completed requested barcode item is added to your cart? because i don't see the item is added to the cart
Note: even i am getting the substring in code it is successfully getting substring but after executing the code, In the POS all the values are getting instead only barcode should display over there

**Reply by madyke** [reply]
Hello - yes, I am using Commerce SDK (version 9.42.22349.7 to be specific, though earlier versions should work as well). And yes, the product is added to the cart.

When you're debugging your extension code, when you're calling `this.defaultExecuteAsync(request)` on line 34, does the `request` object contain a field `scanText` that has the 13-character substring? Or is `scanText` still the full example 24-character barcode?

**Reply by PruthviIntwoC** [reply]
Hi Madyke, 
Yes scanText field has the 13 character substring at line 34 but while running forward in POS it is getting all the full characters.
And
We are using the Retail SDK version 10.0.1227.82, Can you please confirm this is working in Retail SDK as well?

**Reply by madyke** [reply]
Yes, we have confirmed that it is working as expected for Retail SDK. Though again, Retail SDK has been deprecated and due to limited bandwidth are no longer able to support dev issues for Retail SDK, and recommend customers migrate to the new Commerce SDK for extension development.

**Reply by madyke** [reply]
Would it be possible for you to provide a screen recording or a gif (like the one I posted previously) that shows what you're doing as you debug, where the 13-character substring is being selected but then the full barcode is passed to POS?

This is one of our most commonly used extension points, it's unlikely there's an actual issue with the extension point. At this point it's likely some small configuration difference or setup on your extension that is causing the substring to be ignored.

**Reply by PruthviIntwoC** [reply]
Sure, I will share the screen recording, Please give me sometime, thanks

**Reply by PruthviIntwoC** [reply]
Hi madyke, we are having some issue in POS installation in our dev, so can't able to share you screen recording, we will share ASAP. Thanks

**Reply by PruthviIntwoC** [reply]
Hi MadyKe,
Sorry for the delay in response, i got a screen record of the issue which we are facing kindly have a look at it and suggest on the same

https://user-images.githubusercontent.com/119489652/211328909-eb29d608-67e9-4709-be20-88e9ef8dbe4c.mp4

**Reply by madyke** [reply]
Hi Pruthvi, sorry but I'm still not sure what would be going wrong here. At this point our best suggestion would be to switching to using Store Commerce (MPOS is being deprecated in October) and Commerce SDK (Retail SDK is already deprecated and no longer receiving support for issues).

**Reply by dimontei** [reply]
Hi @madyke and @PruthviIntwoC. based in @PruthviIntwoC recording, seems that you don't have any barcode for that Scan text, since mpos is trying to search the values scanned, so this change in GetScanResultClientReqwestHanler will work when you have a barcode setup for an item, please see below:

I just created a valid barcode with 13chars existing in DB + 3 chars 999, that the sample will remove to prove that is working, like:
![image](https://user-images.githubusercontent.com/62382311/213570574-fa7e2873-39cb-4537-9555-499df0fbd31c.png)
 
Barcode setup, will add item 0140:
 
![image](https://user-images.githubusercontent.com/62382311/213570660-48d64333-ed1c-4b79-8299-7d2d8d8fac25.png)

As you can see in the watch section, the sample code is getting the first 13 chars:
![image](https://user-images.githubusercontent.com/62382311/213570744-5cc50c75-aac2-4bb5-9aea-f75374c7e86a.png)

And the result is the item added to the cart:
![image](https://user-images.githubusercontent.com/62382311/213570834-402a0864-c838-445d-b7f6-6dea12c34b45.png)


However if I scan an item without barcode setup, POS will search by the original text 
![image](https://user-images.githubusercontent.com/62382311/213570897-885064d5-f19e-482d-b5a7-9c741b5edf12.png)

So the sample code is working as expected since during the **scan** you can change the "ScanText" value, to search a barcode.
I guess that what you are requesting is the Search by text should get the first 13 chars as well, instead of the original scanned text, when no bar code is found.

Please keep in mind that we also have a CRT Request Handler regarding the Get Scan you may want to check to do your extension business logic:

https://learn.microsoft.com/en-us/dynamics365/commerce/dev-itpro/async-commerce-extension#override-an-out-of-box-request-handler-asynchronously
![image](https://user-images.githubusercontent.com/62382311/213571243-bd63aed5-753b-46bf-ad2d-eab33120593b.png)

**Comment by Danialddm** [comment]
Hi dear
How can I use the output barcode in another handler like getting a serial number?
Actually, I want to pass the param from Getscanrequest to GetserialNum class.
I would be so grateful if guide me.
Regards
Danial

**Reply by madyke** [reply]
Hi @Danialddm - can you please post this as a [new discussion thread on this repo](https://github.com/microsoft/Dynamics365Commerce.Solutions/discussions/new?category=q-a)? We have some internal tracking we're trying to set up and it will be easier to make sure we answer each question if they're separated into their own threads. In your new discussion you can add a link to this thread and explain what you're trying to do there. Thanks!

**Comment by Danialddm** [comment]
done.

On Wed, Jun 7, 2023 at 2:49 AM Matt Dyke ***@***.***> wrote:

> Hi @Danialddm <https://github.com/Danialddm> - can you please post this
> as a new discussion thread on this repo
> <https://github.com/microsoft/Dynamics365Commerce.Solutions/discussions/new?category=q-a>?
> We have some internal tracking we're trying to set up and it will be easier
> to make sure we answer each question if they're separated into their own
> threads. In your new discussion you can add a link to this thread and
> explain what you're trying to do there. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/Dynamics365Commerce.Solutions/discussions/19#discussioncomment-6105282>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXH4AAJF32CVOSCF4D3CT43XJ6UH5ANCNFSM6AAAAAASPS2ZLA>
> .
> You are receiving this because you were mentioned.Message ID:
> <microsoft/Dynamics365Commerce.
> ***@***.***>
>


-- 
*Danial *
POS/MPOS Retail D365FO/AX Developer
